# Here we use One Hot Encoding - variables with many categories:


In [1]:
%pwd

'C:\\Users\\Rose Garden Inn\\Feature Engineering'

In [10]:
#https://www.kaggle.com/aditya1702/mercedes-benz-data-exploration/data

import pandas as pd
import numpy as np



df = pd.read_csv('C:\\Users\\Rose Garden Inn\\Feature Engineering\\mercedesbenz.csv',usecols=['X1','X2','X3','X4','X5'])
df.head()

X1  X2 X3 X4 X5
0  v  at  a  d  u
1  t  av  e  d  y
2  w   n  c  d  x
3  t   n  f  d  x
4  v   n  f  d  h

In [17]:
for x in df.columns:
    print('The Column {} has {} no of categories'.format(x,len(df[x].unique())))

The Column X1 has 27 no of categories
The Column X2 has 44 no of categories
The Column X3 has 7 no of categories
The Column X4 has 4 no of categories
The Column X5 has 29 no of categories


In [18]:
tempdf=df.copy()

In [21]:
pd.get_dummies(tempdf)
tempdf.shape

(4209, 111)

We can observe that from with just 5 categorical features we are getting 111 features with the help of one hot encoding.which is not advisable its called as dimensional curse.

To Come over such problem we can take refernce from 
#### KDD Cup Orange Challenge

What can we do instead?

http://proceedings.mlr.press/v7/niculescu09/niculescu09.pdf
In the winning solution of the KDD 2009 cup: "Winning the KDD Cup Orange Challenge with Ensemble 

###### The Team suggested using 10 most frequent labels convert them into dummy variables using one hot encoding:

In [24]:
# let's find the top 10 most frequent categories for the variable X2
df['X2'].value_counts(ascending=False)[:10]


as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
Name: X2, dtype: int64

In [26]:
top_1oX2=df['X2'].value_counts(ascending=False)[:10].keys()
top_1oX2

Index(['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e'], dtype='object')

In [94]:
top10list=[df[x].value_counts(ascending=False)[:10].keys() for x in df.columns]
top10list
top_1oX1,top_1oX2,top_1oX3,top_1oX4,top_1oX5=top10list
top10list

[Index(['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o'], dtype='object'),
 Index(['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e'], dtype='object'),
 Index(['c', 'f', 'a', 'd', 'g', 'e', 'b'], dtype='object'),
 Index(['d', 'a', 'b', 'c'], dtype='object'),
 Index(['v', 'w', 'q', 'r', 's', 'd', 'n', 'p', 'm', 'i'], dtype='object')]

In [96]:
def one_hot_encoding_top_x1(data,column):
    for cate in top_1oX1:
            data[column+' '+cate]=np.where(data[column]==cate,1,0)
def one_hot_encoding_top_x2(data,column):
    for cate in top_1oX2:
            data[column+' '+cate]=np.where(data[column]==cate,1,0)
def one_hot_encoding_top_x3(data,column):
    for cate in top_1oX3:
            data[column+' '+cate]=np.where(data[column]==cate,1,0)
def one_hot_encoding_top_x4(data,column):
    for cate in top_1oX4:
            data[column+' '+cate]=np.where(data[column]==cate,1,0)
def one_hot_encoding_top_x5(data,column):
    for cate in top_1oX5:
            data[column+' '+cate]=np.where(data[column]==cate,1,0)            
            

In [97]:
one_hot_encoding_top_x1(df,'X1')
one_hot_encoding_top_x2(df,'X2')
one_hot_encoding_top_x3(df,'X3')
one_hot_encoding_top_x4(df,'X4')
one_hot_encoding_top_x5(df,'X5')


In [99]:
df.drop(columns=['X1','X2','X3','X4','X5'],inplace=True)

In [101]:
pd.set_option('display.max_columns', None)
df


X1 aa  X1 s  X1 b  X1 l  X1 v  X1 r  X1 i  X1 a  X1 c  X1 o  X2 as  \
0         0     0     0     0     1     0     0     0     0     0      0   
1         0     0     0     0     0     0     0     0     0     0      0   
2         0     0     0     0     0     0     0     0     0     0      0   
3         0     0     0     0     0     0     0     0     0     0      0   
4         0     0     0     0     1     0     0     0     0     0      0   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...   
4204      0     1     0     0     0     0     0     0     0     0      1   
4205      0     0     0     0     0     0     0     0     0     1      0   
4206      0     0     0     0     1     0     0     0     0     0      0   
4207      0     0     0     0     0     1     0     0     0     0      0   
4208      0     0     0     0     0     1     0     0     0     0      0   

      X2 ae  X2 ai  X2 m  X2 ak  X2 r  X2 n  X2 s  X2 f  X2 e  X3 c  X3 f  \
0         0      0     0      0     0     0     0     0     0     0     0   
1         0      0     0      0     0     0     0     0     0     0     0   
2         0      0     0      0     0     1     0     0     0     1     0   
3         0      0     0      0     0     1     0     0     0     0     1   
4         0      0     0      0     0     1     0     0     0     0     1   
...     ...    ...   ...    ...   ...   ...   ...   ...   ...   ...   ...   
4204      0      0     0      0     0     0     0     0     0     1     0   
4205      0      0     0      0     0     0     0     0     0     0     0   
4206      0      0     0      0     1     0     0     0     0     0     0   
4207      0      0     0      0     0     0     0     0     1     0     1   
4208      1      0     0      0     0     0     0     0     0     1     0   

      X3 a  X3 d  X3 g  X3 e  X3 b  X4 d  X4 a  X4 b  X4 c  X5 v  X5 w  X5 q  \
0        1     0     0     0     0     1     0     0     0     0     0     0   
1        0     0     0     1     0     1     0     0     0     0     0     0   
2        0     0     0     0     0     1     0     0     0     0     0     0   
3        0     0     0     0     0     1     0     0     0     0     0     0   
4        0     0     0     0     0     1     0     0     0     0     0     0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
4204     0     0     0     0     0     1     0     0     0     0     0     0   
4205     0     1     0     0     0     1     0     0     0     0     0     0   
4206     1     0     0     0     0     1     0     0     0     0     0     0   
4207     0     0     0     0     0     1     0     0     0     0     0     0   
4208     0     0     0     0     0     1     0     0     0     0     0     0   

      X5 r  X5 s  X5 d  X5 n  X5 p  X5 m  X5 i  
0        0     0     0     0     0     0     0  
1        0     0     0     0     0     0     0  
2        0     0     0     0     0     0     0  
3        0     0     0     0     0     0     0  
4        0     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0     0  
4205     0     0     0     0     0     0     0  
4206     0     0     0     0     0     0     0  
4207     0     0     0     0     0     0     0  
4208     0     0     0     0     0     0     0  

[4209 rows x 41 columns]